src/sage/algebras/weyl_algebra.py

```diff
@@ -628,17 +629,21 @@ def __classcall__(cls, R, names=None):
             sage: W1 is W2
             True
         """
+        if n is PlusInfinity():  # hook for Infinite weyl algebra
+            return InfiniteDifferentialWeylAlgebra(R, names)
         if isinstance(R, (PolynomialRing_generic, MPolynomialRing_base)):
             if names is None:
                 names = R.variable_names()
                 R = R.base_ring()
+        elif isinstance(R, InfinitePolynomialRing_dense):
+            return InfiniteDifferentialWeylAlgebra(R, names)
```

Comment:
```suggestion
        elif isinstance(R, InfinitePolynomialRing_dense) and names is None:
            return InfiniteDifferentialWeylAlgebra(R.base_ring(), R.variable_names())
```
What if the base ring for a finitely generated Weyl algebra is an infinite polynomial ring? (Add a test checking this works as expected.)

Reply:
This seems to work as expected (added some tests to make sure both infinite over finite and finite over infinite work as expected.
One issue I noticed is that the InfinitePolynomialRing class collapses things down, so it would be tricky to define an infinite weyl algebra over and infinite polynomial ring. For instance, here it forgets that `R` is the base ring.
`sage: R.<x> = InfinitePolynomialRing(QQ)`
`sage: R2.<y> = InfinitePolynomialRing(R)`
`sage: R2.base_ring()`
`Rational Field`